doc/design/refactor/multi_device.md

```diff
@@ -0,0 +1,49 @@
+# Design Doc: Execute the Program with Multi Thread
+
+In PaddlePaddle, the user could declare **wich** operators will be
+execute and **how much** threads will be run in Python code.
+
+## Python Interface
+
+We can specify the thread count with `parallel.do(thread_count=N)`,
+the following PaddlePaddle program shows the usage of the Parallel operator:
+
+```python
+import paddle.v2.fluid as fluid
+N=4
+parallel = fluid.parallel()
+
+x = minibatch([10, 20, 30, 40, 50])
+y = var("y")
+label = var(1)
+w = var("w")
+b = var("b")
+
+fluid.split(x, N)
+
+with parallel.do(thread_count=N):
+    y_predict = fluid.fc(input=x, size=1)
+    cost = fluid.layers.square_error_cost(input=y_predict, label=y)
+    avg_cost = fluid.layers.mean(x=cost)
+
+fluid.merge_grad_avg(x, N) # merge gradient with avg/sum/max/min/...
+sgd_optimizer = fluid.optimizer.SGD(learning_rate=0.001)
+
+with parallel.do(thread_count=N):
+    sgd_optimizer.minimize(avg_cost)
+
+```
+
+## Operator Kernel
+
+- We need a global threadpool, and initialize the threadpool when the
```

Comment:
Since the OP could create the Executor (e.g., while OP), the Executor creation needs to be very lightweight. So perhaps the thread pool initialization should relate with Executor.

Reply:
You mean initialize the ThreadPool in Parallel Op, if that I agree with you.

Reply:
I think ThreadPool will be a singleton (only one instance per process), so it could be initialized in the beginning when the application starts, or lazily when it gets used the first time.

Reply:
Sure, lazily is great, :)